File: app/models/sso_provider.go
package models

import "time"

type SSOProvider struct {
	ID                    uint64    `gorm:"column:id;primaryKey" json:"id"`
	Name                  string    `gorm:"column:name" json:"name"`
	DisplayName           string    `gorm:"column:display_name" json:"display_name"`
	Scene                 string    `gorm:"column:scene" json:"scene"`
	Type                  string    `gorm:"column:type" json:"type"`
	Enabled               bool      `gorm:"column:enabled" json:"enabled"`
	Issuer                string    `gorm:"column:issuer" json:"issuer"`
	Audience              string    `gorm:"column:audience" json:"audience"`
	JWTSecret             string    `gorm:"column:jwt_secret" json:"-"`
	JWTSecretRotatedAt    time.Time `gorm:"column:jwt_secret_rotated_at" json:"jwt_secret_rotated_at"`
	DiscoveryURL          string    `gorm:"column:discovery_url" json:"discovery_url"`
	AuthorizationEndpoint string    `gorm:"column:authorization_endpoint" json:"authorization_endpoint"`
	TokenEndpoint         string    `gorm:"column:token_endpoint" json:"token_endpoint"`
	UserinfoEndpoint      string    `gorm:"column:userinfo_endpoint" json:"userinfo_endpoint"`
	JWKSURI               string    `gorm:"column:jwks_uri" json:"jwks_uri"`
	JWKSJSON              string    `gorm:"column:jwks_json" json:"jwks_json"`
	ClientID              string    `gorm:"column:client_id" json:"client_id"`
	ClientSecret          string    `gorm:"column:client_secret" json:"-"`
	ClientSecretRotatedAt time.Time `gorm:"column:client_secret_rotated_at" json:"client_secret_rotated_at"`
	Scope                 string    `gorm:"column:scope" json:"scope"`
	RedirectURI           string    `gorm:"column:redirect_uri" json:"redirect_uri"`
	EnablePKCE            bool      `gorm:"column:enable_pkce" json:"enable_pkce"`
	EnableNonce           bool      `gorm:"column:enable_nonce" json:"enable_nonce"`
	AutoCreate            bool      `gorm:"column:auto_create" json:"auto_create"`
	Icon                  string    `gorm:"column:icon" json:"icon"`
	ButtonColor           string    `gorm:"column:button_color" json:"button_color"`
	DisplayOrder          int       `gorm:"column:display_order" json:"display_order"`
	SAMLEntrypoint        string    `gorm:"column:saml_entrypoint" json:"saml_entrypoint"`
	SAMLEntityID          string    `gorm:"column:saml_entity_id" json:"saml_entity_id"`
	SAMLCertificate       string    `gorm:"column:saml_certificate" json:"saml_certificate"`
	RoleMapping           JSONMap   `gorm:"column:role_mapping;type:jsonb" json:"role_mapping"`
	DataPermissionMapping JSONMap   `gorm:"column:data_permission_mapping;type:jsonb" json:"data_permission_mapping"`
	AuditColumns
	Timestamps
	Remark string `gorm:"column:remark" json:"remark"`
}

func (SSOProvider) TableName() string {
	return "sso_provider"
}

File: MineAdmin-web/src/modules/base/api/ssoProvider.ts
import type { PageList, ResponseStruct } from '#/global'
import type { SensitiveEvidence } from '~/base/utils/sensitiveOperation'

export type SSOProviderType = 'oidc' | 'oauth2' | 'saml'

export interface SSOProviderVo {
  id?: number
  name?: string
  display_name?: string
  scene?: string
  type?: SSOProviderType
  enabled?: boolean
  issuer?: string
  audience?: string
  jwt_secret?: string
  discovery_url?: string
  authorization_endpoint?: string
  token_endpoint?: string
  userinfo_endpoint?: string
  jwks_uri?: string
  jwks_json?: string
  client_id?: string
  client_secret?: string
  scope?: string
  redirect_uri?: string
  enable_pkce?: boolean
  enable_nonce?: boolean
  auto_create?: boolean
  icon?: string
  button_color?: string
  display_order?: number
  saml_entrypoint?: string
  saml_entity_id?: string
  saml_certificate?: string
  role_mapping?: Record<string, any> | null
  data_permission_mapping?: Record<string, any> | null
  remark?: string
  created_at?: string
  updated_at?: string
  [key: string]: any
}

export interface SSOProviderSearchVo {
  name?: string
  display_name?: string
  scene?: string
  type?: string
  enabled?: string | boolean
  page?: number
  page_size?: number
  [key: string]: any
}

export function page(data: SSOProviderSearchVo): Promise<ResponseStruct<PageList<SSOProviderVo>>> {
  return useHttp().get('/admin/sso-provider/list', { params: data })
}

export function create(data: SSOProviderVo, evidence?: SensitiveEvidence): Promise<ResponseStruct<SSOProviderVo>> {
  return useHttp().post('/admin/sso-provider', { ...data, ...evidence })
}

export function save(id: number, data: SSOProviderVo, evidence?: SensitiveEvidence): Promise<ResponseStruct<SSOProviderVo>> {
  return useHttp().put(`/admin/sso-provider/${id}`, { ...data, ...evidence })
}

export function deleteByIds(ids: number[], evidence: SensitiveEvidence): Promise<ResponseStruct<null>> {
  return useHttp().delete('/admin/sso-provider', { data: { ids, ...evidence } })
}
